src/server_manager/web_app/ui_components/outline-per-key-data-limit-dialog.ts

```diff
@@ -0,0 +1,341 @@
+        this.localize('save')}</paper-button>
+          <paper-button @tap=${this.close}>${this.localize('cancel')}</paper-button>
+        </div>
+      </paper-dialog>
+    `;
+  }
+
+  private _queryAs<T extends HTMLElement>(selector: string): T {
+    return this.shadowRoot.querySelector(selector) as T;
+  }
+
+  private get _input(): PaperInputElement {
+    return this._queryAs<PaperInputElement>('#dataLimitInput');
+  }
+
+  private _dataLimitValue() {
+    return Number(this._input.value);
+  }
+
+  private _dataLimitType() {
+    return this._queryAs<PaperDropdownMenuElement>('#unitsDropdown').selectedItemLabel as 'GB' |
+        'MB';
+  }
+
+  private _initialUnit() {
+    return this._activeDataLimit()?.unit || 'GB';
+  }
+
+  private _initialValue() {
+    return this._activeDataLimit()?.value.toString() || '';
+  }
+
+  private _activeDataLimit(): DisplayDataAmount|undefined {
+    // Returns the limit which currently is enforced on this key, or undefined if there is none.
+    return this.key?.dataLimit || this.serverDefaultLimit;
+  }
+
+  private async _setCustomLimitTapped() {
+    this.showMenu = !this.showMenu;
+    if (this.showMenu) {
+      await this.updateComplete;
+      this._input.focus();
+    }
+  }
+
+  private _setSaveButtonDisabledState() {
+    this.enableSave = !this._input.invalid;
+  }
+
+  private _sendSaveEvent() {
+    const change = this._dataLimitChange();
+    const eventName = `${change === Change.REMOVED ? 'Remove' : 'Save'}PerKeyDataLimitRequested`;
+
+    this.dispatchEvent(new CustomEvent(eventName, {
+      detail: {ui: this},
+      // Required for the event to bubble past a shadow DOM boundary
+      bubbles: true,
+      composed: true,
+    }));
+  }
+
+  /**
+   * Calculates what type of change, or none at all, the user made.
+   */
+  private _dataLimitChange(): Change {
+    const keyLimit = this.key?.dataLimit;
+    if (this.showMenu) {
+      if (!keyLimit) {
+        return Change.SET;
+      }
+      const inputLimit = this.inputDataLimit();
+      if (inputLimit.value !== keyLimit.value || inputLimit.unit !== keyLimit.unit) {
+        return Change.SET;
+      }
+      return Change.UNCHANGED;
+    }
+    // If we unchecked, then the key will have an active data limit.
+    if (keyLimit) {
+      return Change.REMOVED;
+    }
+    return Change.UNCHANGED;
+  }
+
+  /**
+   * Returns true if the data limit was changed by the user.
+   */
+  public dataLimitChanged() {
+    return this._dataLimitChange() !== Change.UNCHANGED;
+  }
+
+  /**
+   * The current data limit as input by the user, but not necessarily as saved.
+   */
+  public inputDataLimit(): DisplayDataAmount {
+    return this.showMenu ? {unit: this._dataLimitType(), value: this._dataLimitValue()} : null;
+  }
+
+  /**
+   * The ID of the key being worked on.  Useful for making API requests for the given key.
+   */
+  public keyId() {
```

Comment:
Where is this used?

Reply:
It's used by `saveAccessKeyDataLimit` and `removeAccessKeyDataLimit` to read which key to operate on from the UI.